lib/truncato/truncated_sax_document.rb

```diff
@@ -2,41 +2,79 @@
 require 'htmlentities'
 
 class TruncatedSaxDocument < Nokogiri::XML::SAX::Document
-  IGNORABLE_TAGS = %w(html head body)
-
-  SINGLE_TAGS = %w{br img}
-
-  attr_reader :truncated_string, :max_length, :max_length_reached, :tail,
-              :count_tags, :filtered_attributes, :filtered_tags, :ignored_levels
+  IGNORABLE_TAGS = %w[html head body].freeze
+
+  # These don't have to be closed (which also impacts ongoing length calculations)
+  SINGLE_TAGS = %w[br img].freeze
+
+  attr_reader :truncated_string,
+    :max_length,
+    :max_length_reached,
+    :tail,
+    :count_tags,
+    :filtered_attributes,
+    :filtered_tags,
+    :ignored_levels,
+    :something_has_been_truncated,
+    :count_bytes
 
   def initialize(options)
     @html_coder = HTMLEntities.new
     capture_options options
     init_parsing_state
   end
 
-  def start_element name, attributes
+  def start_element(name, attributes)
     enter_ignored_level if filtered_tags.include?(name)
     return if @max_length_reached || ignorable_tag?(name) || ignore_mode?
+
+    string_to_add = opening_tag(name, attributes)
+
+    # Abort if there is not enough space to add the combined opening tag and (potentially) the closing tag
+    length = overridden_tag_length(name, string_to_add)
+    return if length > remaining_length
+
+    # Save the tag so we can push it on at the end
     @closing_tags.push name unless single_tag_element? name
-    append_to_truncated_string opening_tag(name, attributes), overriden_tag_length
+
+    append_to_truncated_string string_to_add, length
   end
 
-  def characters decoded_string
-    return if @max_length_reached || ignore_mode?
-    remaining_length = max_length - @estimated_length - 1
-    string_to_append = decoded_string.length > remaining_length ? truncate_string(decoded_string, remaining_length) : decoded_string
-    append_to_truncated_string @html_coder.encode(string_to_append), string_to_append.length
+  def characters(decoded_string)
+    if @max_length_reached || ignore_mode?
+      @something_has_been_truncated |= @max_length_reached
+      return
+    end
+
+    string_to_check = count_bytes ? @html_coder.encode(decoded_string) : decoded_string
+    exceeded_length = char_or_byte_count(string_to_check) > remaining_length
+
+    string_to_append = if exceeded_length
+      @something_has_been_truncated = true
+      truncate_string(decoded_string, remaining_length)
+    else
+      decoded_string
+    end
+    encoded_string_to_append = @html_coder.encode(string_to_append)
+
+    length = if count_bytes
+      # Use encoded length, so &gt; counts as 4 bytes, not 1 (which is what '>' would give)
+      char_or_byte_count(encoded_string_to_append)
+    else
+      # We're counting characters, so don't worry how many bytes it takes to reposent this in the string
+      char_or_byte_count(string_to_append)
+    end
+    append_to_truncated_string encoded_string_to_append, length
```

Comment:
this seems like it would be different behavior than the original, which would pass `string_to_append.length` even though the string itself was encoded..